Add contributing section on code review (#10553)
Qiskit/qiskit: CONTRIBUTING.md

   deprecation, removal etc) that you have added a reno release note for that
   change and that the PR is tagged for the changelog.

### Code review

> All code merged to Qiskit, even from maintainers, goes through a code-review

Reviewer: Same in American English

>    If your code fails the local style checks (specifically the black
>    or Rust code formatting check) you can use `tox -eblack` and

Reviewer 2: Links, in case contributors are not aware of them:

```suggestion
   If your code fails the local style checks (specifically the [black](https://black.readthedocs.io/en/stable/)
   or [Rust code formatting check](https://rust-lang.github.io/rustfmt/)) you can use `tox -eblack` and
```

   `cargo fmt` to automatically fix the code formatting.

2. The documentation has been updated accordingly. In particular, if a
   function or class has been modified during the PR, please update the
   *docstring* accordingly.

   If your pull request is adding a new class, function, or module that is
   intended to be user facing ensure that you've also added those to a
   documentation `autosummary` index to include it in the api documentation.

> 3. If you are of the opinion that the modifications you made warrant additional tests,

Reviewer 2: Less "optional" language?
```suggestion
3. If the modifications you made warrant additional tests,
```

>    feel free to include them.

Reviewer 2: Less "optional" language?

```suggestion
   please include include them.
```

4. Ensure that if your change has an end user facing impact (new feature,

>    deprecation, removal etc) that you have added a `reno` release note for that

Reviewer 2: trying to give examples beyond API changes

```suggestion
   deprecation, removal, bugfix, etc) that you have added a `reno` release note for that
```

>    change and that the PR is tagged for the changelog.

Reviewer 2: Give a bit more importance to the PR changelog labeling
```suggestion
   change.
   
   5. Make sure that the PR gets tagged with `Changelog: X` kind of label. If you are not sure which label you should use, leave it as a comment in your PR. 
```

5. All contributors have [signed the CLA](#contributor-licensing-agreement).

   You will need to ensure that all commits in the PR chain have a correctly configured
   email address, and the email address is registered to a GitHub account that has signed
   the CLA.  A bot will leave a comment with a link to sign the CLA.

6. The PR has a concise and explanatory title that can be understood without
   clicking on another GitHub issue.

   The PR title will become the summary line of the commit, which appears in `git log`.
   For example, "Fixes Issue 1234" is a bad title, and "Fix `ApplyLayout` with
   empty layouts" is good.

7. If the PR addresses an open issue the PR description includes the `fixes #issue-number`

>    syntax to link the PR to that issue (**you must use the exact phrasing in order for GitHub
>    to automatically close the issue when the PR merges**)

Reviewer 2: ```suggestion
   syntax to link the PR to that issue (**you must use [specific phrasing](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) in order for GitHub
   to automatically close the issue when the PR merges**)
```

Remember that spamming issues or pull requests with AI-generated comments is prohibited under the [Qiskit Code of Conduct](https://qisk.it/coc).

### Code review

All code merged to Qiskit, even from maintainers, goes through a code-review
process after a pull request is made.  There are a small number of

> maintainers who can authorize a final merge, but code review involves everyone
> working together to make Qiskit better.  You can review code even if you

Reviewer 2: ```suggestion
maintainers who can authorize a final merge, but **code review involves everyone
working together to make Qiskit better**.  You can review code even if you
```

are not a maintainer, which helps make sure pull requests are technically
correct, well tested, and easier to tackle in their final maintainer review.

> The code-review process is a normal part of software development, and nothing to
> be scared of; for very easy changes it can be as simple as a maintainer saying
> "looks good to me!" (or in short, "LGTM!") and merging the PR.  For more complex changes, it's often a

Reviewer 2: Nothing scares me as somebody saying "nothing to be scared of" `:P`
```suggestion
The code review process is a normal part of software development and is often simpler than expected. For very small changes, it can be as straightforward as a maintainer saying "looks good to me!" (or in short, "LGTM!") and merging the PR. For more complex changes, it's often a
```

back-and-forth where the reviewer may ask a couple of questions about why things
were done a particular way, and make suggestions for improvement.  You don't
need to do everything suggested if you've got good reasons to disagree, but
communicate that clearly and politely.

If you're struggling with code review or a PR on Qiskit, you can ask for help in
the `#qiskit-pr-help` channel on [the public Qiskit Slack](https://qisk.it/join-slack).

Remember that the PR author is a human, not just a username!  It's OK to ask
questions about the code, but don't be mean or rude about it even if you don't
like it.  It's also fine to provide comments that are just compliments with no
suggested changes, if you particularly like something!

#### What to focus on in review

* Is everything in [the PR checklist](#pull-request-checklist) done?

* Are any new public APIs easy to use, well documented, and consistent with
  other parts of the Qiskit API?

* Do any changes to the code have knock-on effects for other parts of Qiskit
  that may be using them, or do they imply changes to the assumptions in

>   our data structures?

Reviewer 2: ```suggestion
  existing data structures?
```

* Are there any edge cases you can think of that the code might not handle well?
  Could the PR benefit from extra tests to cover these, or to verify other edge
  cases that it *does* handle successfully?

* Is the code reasonably easy for you to understand?  This particular point is
  tricky; the more you review code, the easier it will be for you to understand
  other code, so don't worry about this as much if you're getting started.

> * If this PR is a bugfix, is it suitable for backport?  If not, could the PR be

Reviewer 2: Maybe it helps to understand this point (and even the implication of backporting) if:
```suggestion
* If this PR is a bugfix, is it suitable for backport to supported stable versions?  If not, could the PR be
```

  split into a "simple" bugfix that is suitable for backport and a follow-on
  improvement?  (Not all bugfixes *must* be backported.)

#### Writing review comments

* Make concrete suggestions when you think something should be changed, but
  remember that the author might have already thought about it and have a

>   reason.  Try "What do you think about us raising a `TypeError` here instead of

Reviewer 2: sometimes, the use of "us" feels like a counter to "them". Probably the distinction is pointless in this sentence anyway:
```suggestion
  reason.  For example, try "What do you think about raising a `TypeError` here instead of
```

  thoroughly in one go; it's easier to catch bugs like this, and less
  frustrating for the PR creator. If that's too much for you, consider
  asking if the PR could be split into smaller independent chunks.

* Try to keep the number of comments reasonable.  This depends on the size of
  the PR, but remember that there's somebody who'll read all your comments, and
  it can be demoralizing if you get a PR back and it's got 30 comments on from a
  50-line change.  If you feel like you're putting too many comments on,
  consider if you could group several of them into one theme, and ask them as a
  more detailed question with a focus on only one part of the code.  Try not to
  comment the same thing in many places.

* Try to avoid saying "you do" in review comments, and instead try to
  say things like "we do" even when talking about new code. It's not a big
  change, yet it helps to make us think about Qiskit's code as something that we
  all own and care about, and that we're all working together to make it better.

#### Responding to review comments

* Ask questions if you don't understand what a reviewer is saying, or if you're
  not certain whether they're suggesting changes.

> * Feel free to respond to suggestions or questions with your reasoning for doing
>   things a different way, if you don't fully agree with the review comment.

Reviewer 2: "feel free" sounds _very_ optional to me.

```suggestion
* Consider responding to suggestions or questions with your reasoning for doing
  things a different way, if you don't fully agree with the review comment.
```

  particular release (such as a change only to a test), you can leave this
  blank.

  This metadata lets us quickly jump from `git log` to the PR page, and see
  there which Qiskit release a patch went out in.

* The correct "Changelog: X" label is applied, including "Changelog: None" if
  the PR need not appear.

  These labels are much simpler than the `reno` structure; they are for the
  GitHub "releases" page instead, and categorize PRs into "Added", "Deprecated",
  "Changed" or "Fixed".

* Suitable backport commands have been set, if necessary.

  In most cases, applying the label "stable backport potential" is sufficient.
  In this case, the Mergify bot will open a backport PR to the most recent
  stable branch (for example `stable/2.3` if we are currently preparing for
  2.4.0).  If you need more complex backports, write a GitHub comment of the
  form:

  ```
  @Mergifyio backport <branch> <branch2> ...
  ```

  You can have as many branches as necessary.  It usually only necessary to do
  this to support old major branches.

* Any issues fixed by the PR have their own "Fix #<num>" line in the author's PR

>   comment.  If you are empowered to merge PRs, you should be empowered to edit
>   the author's comment to add these, if necessary.

Reviewer 2: No need to edit the original post:

```suggestion
  comment.  If you are empowered to merge PRs, you should be empowered to [link
  the PR to the issue manually](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue-using-the-pull-request-sidebar), if necessary.
```
